File: src/damsindistress.py
n, w = map(int, input().split())
res = []
res.append(w)
for i in range(1, n+1):
    d, c, u = map(int, input().split())
    res.append(max(res[d] - u, c - u))
    w = min(w, res[i])
print(w)

File: src/damsindistress.cpp
#include<bits/stdc++.h>
using namespace std;

int main(){
    int res[200001];
    int n, w; cin >> n >> w;
    res[0] = w;
    for(int i = 1; i <= n; i++){
        int d, c, u;
        cin >> d >> c >> u;
        res[i] = max(res[d]-u, c - u);
        w = min(w, res[i]);
    }
    cout << w << endl;
    return 0;
}